Revamp walkthrough (roughly halfway) (#1386)
argoproj-labs/hera: docs/walk-through/dags.md

# DAGs

> DAGs (Directed Acyclic Graphs) are similar to `Steps`, and the syntax to use DAGs is exactly the same. DAGs are formed

Reviewer: nit: I'd say "very similar" instead of "exactly the same" now you've not got "using context managers" in the sentence, because I'd consider the `>>` to be part of "the syntax" of DAGs.

Author: Fair point, changed some language and updated the paragraph to flow better.

> DAGs (Directed Acyclic Graphs) are similar to `Steps`, and also use a context manager in the same way. DAGs are formed
> of `Tasks`, and offer more flexibility in Workflow construction, with the key difference being that you can specify the

Reviewer: I found the ordering a little confusing: it wasn't immediately obvious that you were still comparing DAGs to Steps by the time you got to "the key difference". I also thought "can specify" might lead people to think it's optional. How about this?

```suggestion
DAGs (Directed Acyclic Graphs) are collections of `Tasks`. They are similar to `Steps`, and use a context manager in the
same way, but offer more flexibility in Workflow construction, with the key difference being that you must specify the
```

Author: Thanks, rejigged the paragraph a bit

          - name: C
            depends: A
            template: echo
            arguments:
              parameters:
              - name: message
                value: C
          - name: D
            depends: B && C
            template: echo
            arguments:
              parameters:
              - name: message
                value: D
      - name: echo
        inputs:
          parameters:
          - name: message
        script:
          image: python:3.9
          source: |-
            import os
            import sys
            sys.path.append(os.getcwd())
            import json
            try: message = json.loads(r'''{{inputs.parameters.message}}''')
            except: message = r'''{{inputs.parameters.message}}'''

            print(message)
          command:
          - python

    ```

The rshift syntax (`>>`) is used with the returned objects from the `echo` calls; it specifies the left-hand-side of the
operator is a dependency of the right-hand-side, i.e. `A >> B` means "B depends on A". This is syntactic sugar for
`A.next(B)`, see the [Task dependencies explained](#task-dependencies-explained) section for more examples. A list acts

> as a boolean `and` of all the elements of the list, and it is important to note that a list cannot appear first in the

Reviewer: maybe put in "(due to Python language limitations)" so people aren't left wondering why we don't support that?
